File: 14-TypeScript/Grey Multi-Tenant/packages/grey-swift/Sources/GreySDK/Domain/MutationClient.swift
// =============================================================================
// Grey Multi-Tenant SDK - Swift
// Domain/MutationClient.swift
//
// Domain client for generic mutation operations.
// Wraps the gRPC service with a clean domain API.
// =============================================================================

import Foundation

// MARK: - Domain Models

/// Mutation result data exposed to consumers.
public struct MutationData: Sendable {
    public let data: Any?
    public let metadata: [String: Any]?
    
    public init(data: Any?, metadata: [String: Any]?) {
        self.data = data
        self.metadata = metadata
    }
}

// MARK: - Client

/// Domain client for generic mutation operations.
/// All methods are async functions that throw GreyError on failure.
public final class MutationClient: Sendable {
    
    private let grpcService: MutationGRPCService
    private let authClient: AuthClient
    
    /// Valid HTTP methods for mutations.
    private static let validMethods: Set<String> = ["POST", "PUT", "PATCH", "DELETE"]
    
    // MARK: - Initialization
    
    internal init(grpcService: MutationGRPCService, authClient: AuthClient) {
        self.grpcService = grpcService
        self.authClient = authClient
    }
    
    // MARK: - Operations
    
    /// Execute a generic mutation.
    /// Requires authentication.
    /// - Parameters:
    ///   - endpoint: The mutation endpoint
    ///   - method: HTTP method (POST, PUT, PATCH, DELETE)
    ///   - body: Optional request body
    /// - Returns: MutationData on success
    /// - Throws: GreyError on failure
    public func mutate(
        endpoint: String,
        method: String = "POST",
        body: Any? = nil
    ) async throws -> MutationData {
        guard !endpoint.isEmpty else {
            throw GreyError.invalidArgument("Endpoint is required")
        }
        
        let normalizedMethod = method.uppercased()
        guard Self.validMethods.contains(normalizedMethod) else {
            throw GreyError.invalidArgument("Invalid method: \(method). Must be one of: \(Self.validMethods)")
        }
        
        guard let accessToken = authClient.accessToken else {
            throw GreyError.unauthenticated()
        }
        
        let response = try await grpcService.mutate(
            accessToken: accessToken,
            endpoint: endpoint,
            method: normalizedMethod,
            body: body
        )
        
        return MutationData(
            data: response.data,
            metadata: response.metadata
        )
    }
    
    /// Execute a POST mutation.
    /// Shorthand for mutate(endpoint:method:body:) with method = "POST".
    /// - Parameters:
    ///   - endpoint: The mutation endpoint
    ///   - body: Optional request body
    /// - Returns: MutationData on success
    /// - Throws: GreyError on failure
    public func post(endpoint: String, body: Any? = nil) async throws -> MutationData {
        try await mutate(endpoint: endpoint, method: "POST", body: body)
    }
    
    /// Execute a PUT mutation.
    /// Shorthand for mutate(endpoint:method:body:) with method = "PUT".
    /// - Parameters:
    ///   - endpoint: The mutation endpoint
    ///   - body: Optional request body
    /// - Returns: MutationData on success
    /// - Throws: GreyError on failure
    public func put(endpoint: String, body: Any? = nil) async throws -> MutationData {
        try await mutate(endpoint: endpoint, method: "PUT", body: body)
    }
    
    /// Execute a PATCH mutation.
    /// Shorthand for mutate(endpoint:method:body:) with method = "PATCH".
    /// - Parameters:
    ///   - endpoint: The mutation endpoint
    ///   - body: Optional request body
    /// - Returns: MutationData on success
    /// - Throws: GreyError on failure
    public func patch(endpoint: String, body: Any? = nil) async throws -> MutationData {
        try await mutate(endpoint: endpoint, method: "PATCH", body: body)
    }
    
    /// Execute a DELETE mutation.
    /// Shorthand for mutate(endpoint:method:body:) with method = "DELETE".
    /// - Parameter endpoint: The mutation endpoint
    /// - Returns: MutationData on success
    /// - Throws: GreyError on failure
    public func delete(endpoint: String) async throws -> MutationData {
        try await mutate(endpoint: endpoint, method: "DELETE", body: nil)
    }
}

File: 14-TypeScript/Grey Multi-Tenant/packages/grey-kotlin/src/main/kotlin/com/grey/sdk/domain/MutationClient.kt
// =============================================================================
// Grey Multi-Tenant SDK - Kotlin
// domain/MutationClient.kt
//
// Domain client for generic mutation operations.
// Wraps the gRPC service with a clean domain API.
// =============================================================================

package com.grey.sdk.domain

import com.grey.sdk.error.GreyError
import com.grey.sdk.error.GreyResult
import com.grey.sdk.grpc.MutationGrpcService
import com.grey.sdk.grpc.MutationResponse

/**
 * Mutation result data exposed to consumers.
 */
data class MutationData(
    val data: Any?,
    val metadata: Map<String, Any>? = null
)

/**
 * Domain client for generic mutation operations.
 * All methods are suspend functions that return GreyResult.
 *
 * @property grpcService The underlying gRPC service
 * @property authClient The auth client for obtaining access tokens
 */
class MutationClient internal constructor(
    private val grpcService: MutationGrpcService,
    private val authClient: AuthClient
) {
    /**
     * Execute a generic mutation.
     * Requires authentication.
     *
     * @param endpoint The mutation endpoint
     * @param method HTTP method (POST, PUT, PATCH, DELETE)
     * @param body Optional request body
     * @return GreyResult containing MutationData or error
     */
    suspend fun mutate(
        endpoint: String,
        method: String = "POST",
        body: Any? = null
    ): GreyResult<MutationData> {
        if (endpoint.isBlank()) {
            return GreyResult.failure(GreyError.invalidArgument("Endpoint is required"))
        }

        val validMethods = setOf("POST", "PUT", "PATCH", "DELETE")
        val normalizedMethod = method.uppercase()
        if (normalizedMethod !in validMethods) {
            return GreyResult.failure(
                GreyError.invalidArgument("Invalid method: $method. Must be one of: $validMethods")
            )
        }

        val accessToken = authClient.accessToken
            ?: return GreyResult.failure(GreyError.unauthenticated())

        return grpcService.mutate(accessToken, endpoint, normalizedMethod, body)
            .map { response -> response.toMutationData() }
    }

    /**
     * Execute a POST mutation.
     * Shorthand for mutate(endpoint, "POST", body).
     *
     * @param endpoint The mutation endpoint
     * @param body Optional request body
     * @return GreyResult containing MutationData or error
     */
    suspend fun post(endpoint: String, body: Any? = null): GreyResult<MutationData> =
        mutate(endpoint, "POST", body)

    /**
     * Execute a PUT mutation.
     * Shorthand for mutate(endpoint, "PUT", body).
     *
     * @param endpoint The mutation endpoint
     * @param body Optional request body
     * @return GreyResult containing MutationData or error
     */
    suspend fun put(endpoint: String, body: Any? = null): GreyResult<MutationData> =
        mutate(endpoint, "PUT", body)

    /**
     * Execute a PATCH mutation.
     * Shorthand for mutate(endpoint, "PATCH", body).
     *
     * @param endpoint The mutation endpoint
     * @param body Optional request body
     * @return GreyResult containing MutationData or error
     */
    suspend fun patch(endpoint: String, body: Any? = null): GreyResult<MutationData> =
        mutate(endpoint, "PATCH", body)

    /**
     * Execute a DELETE mutation.
     * Shorthand for mutate(endpoint, "DELETE", null).
     *
     * @param endpoint The mutation endpoint
     * @return GreyResult containing MutationData or error
     */
    suspend fun delete(endpoint: String): GreyResult<MutationData> =
        mutate(endpoint, "DELETE", null)

    /**
     * Extension to convert gRPC response to domain model.
     */
    private fun MutationResponse.toMutationData(): MutationData = MutationData(
        data = data,
        metadata = metadata
    )
}
